cpu/stm32_common/periph/usbdev.c

```diff
@@ -0,0 +1,1151 @@
+
+/* Some device families forgot to define the FS device FIFO size */
+#ifndef USB_OTG_FS_TOTAL_FIFO_SIZE
+#define USB_OTG_FS_TOTAL_FIFO_SIZE      (1280U)
+#endif
+
+/* Some device families forgot to define the HS device FIFO size */
+#ifndef USB_OTG_HS_TOTAL_FIFO_SIZE
+#define USB_OTG_HS_TOTAL_FIFO_SIZE      (4096U)
+#endif
+
+/* minimum depth of an individual transmit FIFO */
+#define STM32_USB_OTG_FIFO_MIN_WORD_SIZE    (16U)
+/* Offset for OUT endpoints in a shared IN/OUT endpoint bit flag register */
+#define STM32_USB_OTG_REG_EP_OUT_OFFSET      (16U)
+
+/* Endpoint zero size values */
+#define STM32_USB_OTG_EP0_SIZE_64    (0x0)
+#define STM32_USB_OTG_EP0_SIZE_32    (0x1)
+#define STM32_USB_OTG_EP0_SIZE_16    (0x2)
+#define STM32_USB_OTG_EP0_SIZE_8     (0x3)
+
+/* Endpoint type values */
+#define STM32_USB_OTG_EP_TYPE_CONTROL   (0x00 << USB_OTG_DOEPCTL_EPTYP_Pos)
+#define STM32_USB_OTG_EP_TYPE_ISO       (0x01 << USB_OTG_DOEPCTL_EPTYP_Pos)
+#define STM32_USB_OTG_EP_TYPE_BULK      (0x02 << USB_OTG_DOEPCTL_EPTYP_Pos)
+#define STM32_USB_OTG_EP_TYPE_INTERRUPT (0x03 << USB_OTG_DOEPCTL_EPTYP_Pos)
+
+/* List of instantiated USB peripherals */
+static stm32_usb_otg_fshs_t _usbdevs[USBDEV_NUMOF] = { 0 };
+
+static usbdev_ep_t _out[_TOTAL_NUM_ENDPOINTS];
+static usbdev_ep_t _in[_TOTAL_NUM_ENDPOINTS];
+
+/* Forward declaration for the usb device driver */
+const usbdev_driver_t driver;
+
+static void _flush_fifo(const stm32_usb_otg_fshs_config_t *conf,
+                        uint8_t fifo_num);
+
+/*************************************************************************
+* Conversion function from the base address to specific register blocks *
+*************************************************************************/
+static USB_OTG_GlobalTypeDef *_global_regs(
+    const stm32_usb_otg_fshs_config_t *conf)
+{
+    return (USB_OTG_GlobalTypeDef *)(conf->periph + USB_OTG_GLOBAL_BASE);
+}
+
+static USB_OTG_DeviceTypeDef *_device_regs(
+    const stm32_usb_otg_fshs_config_t *conf)
+{
+    return (USB_OTG_DeviceTypeDef *)(conf->periph + USB_OTG_DEVICE_BASE);
+}
+
+static USB_OTG_INEndpointTypeDef *_in_regs(
+    const stm32_usb_otg_fshs_config_t *conf,
+    size_t endpoint)
+{
+    return (USB_OTG_INEndpointTypeDef *)(conf->periph +
+                                         USB_OTG_IN_ENDPOINT_BASE +
+                                         USB_OTG_EP_REG_SIZE * endpoint);
+}
+
+static USB_OTG_OUTEndpointTypeDef *_out_regs(
+    const stm32_usb_otg_fshs_config_t *conf,
+    size_t endpoint)
+{
+    return (USB_OTG_OUTEndpointTypeDef *)(conf->periph +
+                                          USB_OTG_OUT_ENDPOINT_BASE +
+                                          USB_OTG_EP_REG_SIZE * endpoint);
+}
+
+static __I uint32_t *_rx_fifo(const stm32_usb_otg_fshs_config_t *conf)
+{
+    return (__I uint32_t *)(conf->periph + USB_OTG_FIFO_BASE);
+}
+
+static __O uint32_t *_tx_fifo(const stm32_usb_otg_fshs_config_t *conf,
+                              size_t num)
```

Comment:
Why is there an extra param (size_t num) for tx_fifo but not for rx_fifo ?

Reply:
Because there are multiple TX FIFOs (for IN transfers), but only a single shared RX FIFO (for OUT transfers).